### PR TITLE
Add GPT-NeoX decoder and causal language model

### PR DESCRIPTION
**Pull request by danieldk**
## Description

In order to support outputting predicted tokens and key/value caching, two new output classes are added: `ModelOutputWithCache` and `CausalLMOutputWithCache`. The existing `PyTorchTransformerOutput` is renamed to `ModelOutput`.

This change also adds a new `SelfAttentionWithRotaryEmbeddings` layer, which is similar to the existing self-attention layer except that it also applies rotary embeddings.

This change also adds support for causal masking to the self-attention mechanism to support autoregressive models.

### Types of change

New model type.

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: curated_transformers/models/attention.py [hunk 1]
```diff
@@ -143,12 +193,163 @@ def forward(self, x: Tensor, attn_mask: Optional[AttentionMask]) -> Tensor:
...
+        """
+
+        super().__init__()
+
+        self.model_dim = hidden_width
+        self.num_heads = num_attention_heads
+        if self.model_dim % self.num_heads != 0:
+            raise ValueError(
+                f"The hidden width of the transformer ({self.model_dim}) must be "
+                "divisible by the number of self-attention heads ({self.num_heads})"
```
Comment (shadeMe) [inline curated_transformers/models/attention.py:238]:
```suggestion
                f"divisible by the number of self-attention heads ({self.num_heads})"
```

File: curated_transformers/models/output.py [hunk 2]
```diff
@@ -39,3 +47,64 @@ def last_hidden_layer_states(self) -> Tensor:
     @property
     def all_hidden_layer_states(self) -> List[Tensor]:
         return self.all_outputs[1:]
+
+
+class ModelOutputWithCache(Generic[CacheT], ModelOutput):
+    """Output of causal language models.
+
+    :cache: Model cache. The cache can be used by future calls to
```
Comment (shadeMe) [inline curated_transformers/models/output.py:55]:
```suggestion
    :cache: Model cache. The cache can be used with future calls to
```

File: curated_transformers/tests/models/gpt_neox/test_causal_lm.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import pytest
+import torch
+
+from curated_transformers._compat import has_hf_transformers, transformers
+from curated_transformers.models.gpt_neox.causal_lm import GPTNeoXCausalLM
+from curated_transformers.tests.util import torch_assertclose
+
+
+@pytest.mark.slow
```
Comment (shadeMe) [inline curated_transformers/tests/models/gpt_neox/test_causal_lm.py:9]: We can remove the slow markers now, I think?

File: curated_transformers/tests/models/gpt_neox/test_decoder.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
+import pytest
+import torch
+
+from curated_transformers._compat import has_hf_transformers, transformers
+from curated_transformers.models.gpt_neox.decoder import GPTNeoXDecoder
+from curated_transformers.tests.util import torch_assertclose
+
+
+@pytest.mark.skipif(not has_hf_transformers, reason="requires huggingface transformers")
+@pytest.mark.slow
```
Comment (shadeMe) [inline curated_transformers/tests/models/gpt_neox/test_decoder.py:10]: Don't need these markers anymore with the tiny test models.

File: curated_transformers/models/output.py [hunk 5]
```diff
@@ -39,3 +47,64 @@ def last_hidden_layer_states(self) -> Tensor:
...
+    # Shape: [batch_size, seq_len, vocab_size]
+    logits: Tensor
+
+    def __init__(
+        self,
+        *,
+        cache: Optional[List[CacheT]],
+        embedding_output: Tensor,
+        layer_hidden_states: List[Tensor],
+        logits: Tensor,
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/models/output.py:96]: I was starting to order kwarg-only lists alphabetically.

[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/models/output.py:96]: But not consistently in the docs 🙈

File: curated_transformers/models/gpt_neox/_hf.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+def convert_hf_state_dict(cls, params: Mapping[str, Parameter]) -> Mapping[str, Tensor]:
+    """Convert state dict from HF paramater naming to ours.
+    The function is insensitive to prefixes, to allow loading
+    both the decoder and the full LM."""
+    if issubclass(cls, DecoderModule):
+        stripped_params = {
+            re.sub(r"^gpt_neox\.", "", k): v
+            for k, v in params.items()
+            if "embed_out" not in k
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/models/gpt_neox/_hf.py]: Yeah, otherwise the decoder will fail to load from the state dict because there is an unused parameter. I have made the check more explicit and added a comment.

File: curated_transformers/models/gpt_neox/config.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        :param vocab_size: Vocabulary size (number of embeddings).
+        """
+
+        # TODO: max_position_embeddings and model_max_length are currently
+        #       not used. We may want to limit the rotary embeddings to these
+        #       values in the future. We should check empirically if the auto
+        #       resizing in rotary embeddings makes sense.
+
+        self.attention = GPTNeoXAttentionConfig(
+            dropout_prob=0.0,
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/models/gpt_neox/config.py]: Fixed.

File: curated_transformers/models/gpt_neox/decoder.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+            input_ids, attention_mask, token_type_ids - (batch, seq_len)
+        """
+        embeddings = self.embeddings(input_ids)
+        layer_output = embeddings
+
+        layer_outputs = []
+        new_cache = []
+        layer_cache = None
+        for layer in self.layers:
+            if cache is not None:
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/models/gpt_neox/decoder.py:80]: Added a short comment here, but put a longer explanation of caching in the Causal LM docs.

File: curated_transformers/models/attention.py [hunk 9]
```diff
@@ -143,12 +193,163 @@ def forward(self, x: Tensor, attn_mask: Optional[AttentionMask]) -> Tensor:
...
+        :param dropout_prob: Dropout to apply between the self-attention
+            and output layers.
+        :param hidden_width: Hidden width of the layer.
+        :param num_attention_heads: Number of attention heads.
+        :param rotary_fraction: fraction of hidden width to apply rotary
+            embeddings to. Must be in [0,1].
+        :param rotary_base: Base in signifying the rotary embedding period.
+        :param split_heads_before_chunk: Split the hidden representation
+            into heads before splitting query/key/value representations when
+            ``True``.
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/models/attention.py]: Added a short line explaining that this is necessary for some newer architectures. (I don't think there is a real benefit though?)

File: curated_transformers/tests/models/gpt_neox/test_causal_lm.py (hunk 3 above)
Comment (danieldk) [reply curated_transformers/tests/models/gpt_neox/test_causal_lm.py]: Done.

**Review by shadeMe — APPROVED** [review]
